test/test_optim.py

```diff
@@ -640,15 +642,125 @@ def test_reduce_lr_on_plateau8(self):
+
+        diff_lr_1 = max_lr_1 - base_lr_1
+        diff_lr_2 = max_lr_2 - base_lr_2
+
+        gamma = 0.9
+        xs = [0, 0.25, 0.5, 0.75, 1, 0.75, 0.50, 0.25, 0, 0.25, 0.5, 0.75, 1]
+        target_1 = list(map(lambda x: base_lr_1 + x[1] * diff_lr_1 * gamma**x[0], enumerate(xs)))
+        target_2 = list(map(lambda x: base_lr_2 + x[1] * diff_lr_2 * gamma**x[0], enumerate(xs)))
+        targets = [target_1, target_2]
+        scheduler = CyclicLR(self.opt, base_lr=[base_lr_1, base_lr_2],
+                             max_lr=[max_lr_1, max_lr_2], step_size_up=4,
+                             mode='exp_range', gamma=gamma)
+        self._test_cycle_lr(scheduler, targets, len(target_1))
+
+    def test_triangular_mode_step_size_up_down(self):
+        target = [1.0, 2.0, 3.0, 4.0, 5.0, 13.0 / 3, 11.0 / 3, 9.0 / 3, 7.0 / 3, 5.0 / 3, 1.0]
+        targets = [target, target]
+        scheduler = CyclicLR(self.opt, base_lr=1, max_lr=5,
+                             step_size_up=4,
+                             step_size_down=6,
+                             mode='triangular')
+        self._test_cycle_lr(scheduler, targets, len(target))
+
+    def test_triangular2_mode_step_size_up_down(self):
+        base_target = ([
+            1.0, 3.0, 5.0, 13.0 / 3, 11.0 / 3, 9.0 / 3, 7.0 / 3, 5.0 / 3, 1.0, 2.0, 3.0, 8.0 / 3,
+            7.0 / 3, 6.0 / 3, 5.0 / 3, 4.0 / 3, 1.0, 3.0 / 2, 2.0, 11.0 / 6, 10.0 / 6, 9.0 / 6,
+            8.0 / 6, 7.0 / 6
+        ])
+        deltas = [2 * i for i in range(0, 2)]
+        base_lrs = [1 + delta for delta in deltas]
+        max_lrs = [5 + delta for delta in deltas]
+        targets = [[x + delta for x in base_target] for delta in deltas]
+        scheduler = CyclicLR(
+            self.opt,
+            base_lr=base_lrs,
+            max_lr=max_lrs,
+            step_size_up=2,
+            step_size_down=6,
+            mode='triangular2')
+        self._test_cycle_lr(scheduler, targets, len(base_target))
+
+    def test_exp_range_mode_step_size_up_down(self):
+        base_lr, max_lr = 1, 5
+        diff_lr = max_lr - base_lr
+        gamma = 0.9
+        xs = ([
+            0.0, 0.5, 1.0, 5.0 / 6, 4.0 / 6, 3.0 / 6, 2.0 / 6, 1.0 / 6, 0.0, 0.5, 1.0, 5.0 / 6,
+            4.0 / 6
+        ])
+        target = [base_lr + x * diff_lr * gamma**i for i, x in enumerate(xs)]
+        targets = [target, target]
+        scheduler = CyclicLR(self.opt, base_lr=base_lr, max_lr=max_lr,
+                             step_size_up=2, step_size_down=6,
+                             mode='exp_range', gamma=gamma)
+        self._test_cycle_lr(scheduler, targets, len(target))
+
     def test_lambda_lr(self):
         epochs = 10
         self.opt.param_groups[0]['lr'] = 0.05
         self.opt.param_groups[1]['lr'] = 0.4
-        targets = [[0.05 * (0.9 ** x) for x in range(epochs)], [0.4 * (0.8 ** x) for x in range(epochs)]]
+        targets = [[0.05 * (0.9 ** x) for x in range(epochs)], [0.4 * (0.8 ** x)
```

Comment:
please only change lines relevant to your PR.
We have line limits of 120 characters, so your changes here and above in this file are not needed, and actively introduce noise.